src/coreclr/src/vm/assemblyloadcontext.cpp

```diff
@@ -14,3 +15,49 @@ HRESULT AssemblyLoadContext::GetBinderID(
     *pBinderId = reinterpret_cast<UINT_PTR>(this);
     return S_OK;
 }
+
+NativeImage *AssemblyLoadContext::LoadNativeImage(Module *componentModule, LPCUTF8 nativeImageName)
+{
+    STANDARD_VM_CONTRACT;
+
+#ifndef DACCESS_COMPILE
+    BaseDomain::LoadLockHolder lock(AppDomain::GetCurrentDomain());
+    AssemblyLoadContext *loadContext = componentModule->GetFile()->GetAssemblyLoadContext();
+    PTR_LoaderAllocator moduleLoaderAllocator = componentModule->GetLoaderAllocator();
+
+    int nativeImageCount = m_nativeImages.GetCount();
+    for (int nativeImageIndex = 0; nativeImageIndex < nativeImageCount; nativeImageIndex++)
+    {
+        NativeImage *nativeImage = m_nativeImages[nativeImageIndex];
+        if (nativeImage->Matches(nativeImageName))
+        {
+            return nativeImage;
+        }
+    }
+    
+    SString path = componentModule->GetPath();
+    SString::Iterator lastPathSeparatorIter = path.End();
+    size_t pathDirLength = 0;
+    if (PEAssembly::FindLastPathSeparator(path, lastPathSeparatorIter))
+    {
+        pathDirLength = (lastPathSeparatorIter - path.Begin()) + 1;
+    }
+
+    SString compositeImageFileName(SString::Utf8, nativeImageName);
+    SString fullPath;
+    fullPath.Set(path, path.Begin(), (COUNT_T)pathDirLength);
+    fullPath += compositeImageFileName;
+
+    PTR_PEImage peImage = PEImage::OpenImage(fullPath);
+    PTR_PEFile peFile = PEFile::Open(peImage);
```

Comment:
You should not need to open PEFile or PEImage. Just create PEImageLayout without the extra ceremony.

Reply:
I apologize, I'm not getting this - how can I get a PEImageLayout without a PEImage?

Reply:
For Windows, it can look like this:

```
#ifdef TARGET_WINDOWS
    m_hModule = CLRLoadLibraryEx(path,NULL,GetLoadWithAlteredSearchPathFlag());
    if (m_hModule == NULL)
    {
        HRESULT hr = HRESULT_FROM_GetLastError();
        EEFileLoadException::Throw(path, hr, NULL);
    }
    m_pPEDecoder = new PEDecoder((void*)m_hModule, TRUE);
#else
```

For Unix, it will be a bit more complex - the key functions to call are `PAL_LOADLoadPEFile` and `ApplyBaseRelocations()`.

The reason to avoid PEImage is that it has multiple ways to load stuff, and it does fallbacks between different ways for compatibility. You do not need any of that for this: You know that the file should be a good native image, and if it is not - the right thing to do is to fail immediately.

Reply:
Thanks for the explanation. I'm trying to make this change but I guess I'm still missing how to construct a PEImageLayout based on the snippet you kindly provided and switching over all related logic to use PEDecoder seems pretty viral, especially due to <code>Module::GetNativeOrReadyToRunImage</code>. If that is what you're truly suggesting, I'll probably make a preparatory change just "downgrading" these various places to use PEDecoder instead of PEImageLayout to simplify integration of this change.

Reply:
Keeping PEImageLayout is ok - I have not realized that it is viral. PEImage and PEFile are the important ones to avoid.